### DISCUSSION TITLE
Link API medusa v2

### DISCUSSION THREAD

**Discussion by Edsparr** [medusajs/medusa#10108] [Feature Requests]
Are there any plans to provide a more flexible linking api than entity id to entity id?

For example, if I want to create a localized product entity which have a composite primary key of (product id, language code) it makes a lot of sense to link it directly to the product table. Right now that is not possible and a separate id for each localized entity need to be created, then linked.

I understand the reason behind the current API which offers strong flexibility and modularity. Still though, without focusing on the extra table the joins will stack up overtime. Especially in scenarios containing nested joins it can become a lot. Would be great to hear the core team's thoughts around the subject.

**Comment by carlos-r-l-rodrigues** [comment]
Hi @Edsparr,
If you just need to connect 2 entities, but you don't need to manage the relationship between them, because one side already has the reference to the other, you can have read-only links.

something like:
```
defineLink(localizedProduct.linkable.product_id, ProductModule.linkable.product, { readOnly: true })
```

We are still improving our docs to contain all the features and polishing the DX of that.
But read only links are meant for this, and when you query either the product or your localized product, you can expand to fetch the relation.

On the other hand, at the moment it won't be managed by the application in case a product is deleted, it won't cascade delete the entity on your custom entity.

**Reply by Edsparr** [reply]
That's super interesting, thank you for your response Carlos!

And in my case where I would want another field with product_id (language_code) to form a many to one relationship I could just have a composite key instead of a single primary key and your example would work? 

When querying this though, how could I filter by language_code? Is this not possible until the index/search module is released?

**Reply by carlos-r-l-rodrigues** [reply]
Yeah, you can manage the keys as you prefer on your module, just keeping in mind that depending on how you expand you query, the module will be requested only to filter `product_id` or the composition of product_id + lang.

The `query.graph` api let you pass filters to the main entity and also to the relationships. So you can list products, expanding to you custom module filtering by the language.

In this case you don't really need the index module. The query.graph is an aggregator, if you query product, and expand to you module filtering by a language that doesn't exist, it will just return empty.

You would need the Index module in case you need to filter out products that don't have the language you filtered by.

**Reply by Edsparr** [reply]
I will try it out, thanks for the help!

**Reply by Edsparr** [reply]
> Yeah, you can manage the keys as you prefer on your module, just keeping in mind that depending on how you expand you query, the module will be requested only to filter `product_id` or the composition of product_id + lang.
> 
> The `query.graph` api let you pass filters to the main entity and also to the relationships. So you can list products, expanding to you custom module filtering by the language.
> 
> In this case you don't really need the index module. The query.graph is an aggregator, if you query product, and expand to you module filtering by a language that doesn't exist, it will just return empty.
> 
> You would need the Index module in case you need to filter out products that don't have the language you filtered by.

Hi again Carlos! I sat down to play with it and see if I could figure it out but I am having some difficulties. When registering my readonly link the generated typing for the query API disappears. Also when ignoring typing rules it simply does not return anything (this approach has worked before for some reason for normal linking). Would be great if you could provide me with a working example of making a readonly link to an entity with a many to one relationship. If possible also include example of fetching the original entity with the relationship to the linked entity with a filter based on one of the linked entity's field. More concretly, fetching a product. Fetching the localizedProduct entity (which has a product_id language_code composite key) and filter based on localizedProduct's language code.

I think many people could learn from this example too, seems to be some confusion around the link api.

Thank you in advance!

**Reply by p-stalega** [reply]
hi @Edsparr  - any news on this? I'm struggling with the same issue now, having difficulties to filter by linked data - did you figure out a solution?
